pkg/alertmanager/api.go

```diff
@@ -98,13 +100,30 @@ func (am *MultitenantAlertmanager) SetUserConfig(w http.ResponseWriter, r *http.
 		return
 	}
 
-	payload, err := ioutil.ReadAll(r.Body)
+	var input io.Reader
+	maxConfigSize := am.limits.AlertmanagerMaxConfigSize(userID)
+	if maxConfigSize > 0 {
+		// LimitReader will return EOF after reading specified number of bytes. To check if
+		// we have read too many bytes, allow one extra byte.
+		input = io.LimitReader(r.Body, int64(maxConfigSize)+1)
+	} else {
+		input = r.Body
+	}
+
+	payload, err := ioutil.ReadAll(input)
 	if err != nil {
 		level.Error(logger).Log("msg", errReadingConfiguration, "err", err.Error())
 		http.Error(w, fmt.Sprintf("%s: %s", errReadingConfiguration, err.Error()), http.StatusBadRequest)
 		return
 	}
 
+	if maxConfigSize > 0 && len(payload) > maxConfigSize {
+		msg := fmt.Sprintf(errConfigurationTooBig, maxConfigSize)
+		level.Error(logger).Log("msg", msg)
```

Comment:
```suggestion
		level.Error(logger).Log("msg", msg, "user", userID)
```

nit: I noticed none of these logs in this function include the userID, I wouldn't say change them all (but maybe we do 😄 ) but let's start the trend.

Reply:
The reason why logger doesn’t need user is that the following line at the top of this method:

```
logger := util_log.WithContext(r.Context(), am.logger)
```

will already put tenant into logger, if one is present in the context.

Reply:
TIL, all good then.

Reply:
Thanks for your review!